pkg/grpc/server.go

```diff
@@ -147,6 +156,33 @@ func (a *apiImpl) Register(services ...APIService) {
 	a.apiServices = append(a.apiServices, services...)
 }
 
+func (a *apiImpl) Stop() bool {
+	if a.shutdownInProgress.IsDone() {
+		return false
+	}
+
+	if a.stopMutex.TryLock() {
```

Comment:
I think my suggestion was wrong, there does not need to be a mutex because there is nothing to be released - a second call to `Stop()` does result in an error, a `Mutex` would make sense if it necessary to block here until the next caller can access a data structure or method.

Instead of using a Mutex using the atomic package seems more clean to indicate a `Stop` is in progress and disallow multiple calls.

For example:

```
var (
    locker    uint32
    errLocked = errors.New("Locked out buddy")
)

func OneAtATime(d time.Duration) error {
    if !atomic.CompareAndSwapUint32(&locker, 0, 1) { // <-----------------------------
        return errLocked                             //   All logic in these         |
    }                                                //   four lines                 |
    defer atomic.StoreUint32(&locker, 0)             // <-----------------------------

    // logic here, but we will sleep
    time.Sleep(d)

    return nil
}
``` 

The `if` case can check on the `locker` variable instead of listening to a signal.

Source: https://stackoverflow.com/questions/45208536/good-way-to-return-on-locked-mutex-in-go

By using the atomic package the signal and mutex can be reduced into a single variable.

Reply:
Makes sense. I've used `atomic.Bool` instead, and was able to get rid of both the lock and the signal. Let me know what do you think.